doc/lua_api.txt

```diff
@@ -4155,14 +4174,30 @@ Functions receive a "luaentity" as `self`:
 
 Callbacks:
 
-* `on_activate(self, staticdata, dtime_s)`
-    * Called when the object is instantiated.
-    * `dtime_s` is the time passed since the object was unloaded, which can be
-      used for updating the entity state.
-* `on_step(self, dtime)`
+* `on_activate(self, staticdata, dtime_s, id)`
+    * Called immediately after the object is added to the environment.
+    * `dtime_s`: time passed since the object was unloaded, which can be used
+      for updating the entity state.
+    * `id`: active object identifier used internally by the engine and stored
+      in the `minetest.luaentities` table
+    * Note that value of `id` is only guaranteed to be unique for the lifetime 
+      of the object and should never be referenced once deactivated.
+* `on_deactivate(self, id)`
+    * Called immediately before the object is removed from the environment,
+      whether as the result of the mapblock unloading, the object being outside
+      the active-object radius, an explicit call to `remove()`, etc.
+* `on_step(self, dtime, pos, rot, new_vel, old_vel, move_result)`
```

Comment:
1) This PR was not "breaking backwards compatibility" back when it was originally submitted, since the only parameters to the on_step callback were "self" and "dtime".

2) Passing the position, rotation, and velocity as parameters saves repeated context-switches to CPP. Since these same values are often queried on every server step, there is a clear benefit to streamlining this potential bottleneck.

3) "Poor code quality" is a vague and nebulous criticism without substance, very reminiscent of your earlier claim "This adds quite some overhead for non-mob entities," even though non-mob entities are immobile and non-physical.

4) Claiming "a bunch of unnecessary utilities" is of dubious veracity, since every feature of this PR can be proven to have value and merit. This exact API is used by multiple mods on my production server. What server do you operate? 

5) Claiming "new_vel and old_vel are fields of the moveresult" is patently misleading, as I pointed out previously:

https://github.com/minetest/minetest/pull/9717#discussion_r438812372

> The old_velocity and new_velocity are entirely different from the results that this PR provides. The distinction is even important since the moveresult is only available when the object is physical=true. Yet non-physical objects can have a change in velocity. So that is why I am reporting velocity changes separate from the move result.
>
> In addition, the 'old_velocity' is a very misleading name since it's not actually the old velocity, it's the intended velocity once acceleration has been applied AND as a result of each prior collision box processed within the loop. As a result it's not dependent on any specific node/object collision, but rather on an implementation-specific order of collision processing which could change at any time in the future, making it highly volatile for use in modding.
>
> As for new_velocity that is even less intuitive. Just like old_velocity, the value is entirely dependent on the sequence of collision boxes already processed within the loop but also the current node/object collision, so it doesn't reflect anything about the object's "new velocity" that is particularly useful for modders.

The fact you wrote an entire mob-related library, yet do not even understand the basic mechanics of how collisions are processed by the engine is rather concerning.

Reply:
>     1. This PR was not "breaking backwards compatibility" back when it was originally submitted, since the only parameters to the on_step callback were "self" and "dtime".

I know, but right now, it is breaking compatibility, and that's what matters if you want it merged now. 5.3 has shipped.

>     2. Passing the position, rotation, and velocity as parameters saves repeated context-switches to CPP. Since these same values are often queried on every server step, there is a clear benefit to streamlining this potential bottleneck.

Premature optimization. That method header is unnecessarily bloated. Do you know what a primary key is?

>     3. "Poor code quality" is a vague and nebulous criticism without substance, very reminiscent of your earlier claim "This adds quite some overhead for non-mob entities," even though non-mob entities are immobile and non-physical.

Non-mob entities are immobile and non-physical? Falling nodes beg to differ. As celeron pointed out, the overhead is indeed most likely negligible. `old_velocity` and `new_velocity` should definitely be moveresult fields.


>     4. Claiming "a bunch of unnecessary utilities" is of dubious veracity, since every feature of this PR can be proven to have value and merit. This exact API is used by multiple mods on my production server. What server do you operate?

Your server which hasn't even bothered migrating to 5.x?

I was not trying to say that your features "useless", but rather that they are not necessary, and should probably not be part of the engine. They should rather be written in Lua as they are purely serverside anyways and don't require anything which isn't already there; inclusion in builtin could be debated.


> The fact you wrote an entire mob-related library, yet do not even understand the basic mechanics of how collisions are processed by the engine is rather concerning.

The fact that you make such dubious unfounded claims is rather concerning.